python-threatexchange/threatexchange/TE.py

```diff
@@ -13,15 +13,32 @@
 # General Python dependencies
 import os
 import re
-import urllib
-import urllib.error
+import requests
+from requests.packages.urllib3.util.retry import Retry
+
 import urllib.parse
-import urllib.request
+
+from .common import TimeoutHTTPAdapter
+
+
+DEFAULT_TE_BASE_URL = "https://graph.facebook.com/v6.0"
+retry_strategy = Retry(
+    total=4,
+    status_forcelist=[429, 500, 502, 503, 504],
+    method_whitelist=["HEAD", "GET", "OPTIONS"],
+)
+adapter = TimeoutHTTPAdapter(timeout=60, max_retries=retry_strategy)
+
+# `fb_graph_api`: Custom requests session that provides
+# - retries: 4 retries on GET requests for 429 or 5XX error codes
+# - timeout: 60 seconds. Can be adjusted at the fb_graph_api.get(.., timeout=FOO) level
+
+fb_graph_api = requests.Session()
+fb_graph_api.mount(DEFAULT_TE_BASE_URL, adapter=adapter)
```

Comment:
Hmm, in general I would say avoid doing work in the initialization of a module. However, I haven't read enough on requests to see if this is doing any work or if it's just populating fields on an unopened connection. Do you know whether this is doing anything significant?


The alternative would be to have a get_client() method which is memoized.

Reply:
The 'Session' Object would use connection pooling when connecting to the same endpoint.

It is populating the timeout value, but also implementing a retry strategy.

About doing work in module init, the only issue I see is with reusing the same Session object when doing concurrent fetches as we are doing in [FetchCommand](https://github.com/facebook/ThreatExchange/blob/master/python-threatexchange/threatexchange/cli/fetch.py#L204). I'll check if reusing the same Session object is possible/advisable in a ThreadPoolExecutor environment.

FWIW: `requests.get` [creates a new session each time](https://github.com/psf/requests/blob/master/requests/api.py#L60). I'll update the PR with my findings.